### PR TITLE
Data Library Syncs

### PR DESCRIPTION
**Pull request by Kartones**
Closes https://github.com/CartoDB/cartodb/issues/4775
Closes https://github.com/CartoDB/cartodb/issues/5028

### CODE REVIEW

**Comment by Kartones** [comment]
@rafatower CR please

**Comment by javierarce** [comment]
@xavijam: have a look to the frontend code, please.

File: app/models/carto/external_source.rb [hunk 1]
```diff
@@ -3,4 +3,7 @@
 require 'active_record'
 
 class Carto::ExternalSource < ActiveRecord::Base
+
+  REFRESH_INTERVAL = 30*24*60*60
```
Comment (rafatower) [inline app/models/carto/external_source.rb]: units?

File: app/models/synchronization/member.rb [hunk 2]
```diff
@@ -502,6 +502,10 @@ def get_datasource(datasource_name)
         datasource
       end
 
+      def from_external_source?
+        ::ExternalDataImport.where(synchronization_id: self.id).first != nil
```
Comment (rafatower) [inline app/models/synchronization/member.rb]: just wondering if fetching the whole thing is really needed for an "presence test"

**Comment by rafatower** [comment]
:+1: 
sorry for the late review
backend looks good to me

File: app/models/synchronization/member.rb (hunk 2 above)
Comment (Kartones) [inline app/models/synchronization/member.rb]: I can pluck only the Id, but we're doing this all way around the code, and the `ExternalDataImport` model is quite small, has very few fields...

File: lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js [hunk 3]
```diff
@@ -61,7 +68,10 @@ module.exports = cdb.core.Model.extend({
     // Remote?
     if (type === "remote") {
       _.extend(d, {
+        type: "remote",
+        interval: null,
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js]: So, if interval is null, how does import process know that it needs to be synced?

File: lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js [hunk 4]
```diff
@@ -7,7 +7,13 @@ var IntervalView = require('./sync/interval_view');
  */
 module.exports = BaseDialog.extend({
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js]: Caution, this view will have conflicts with this [PR](https://github.com/CartoDB/cartodb/pull/5094). @viddo

File: lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js [hunk 5]
```diff
@@ -91,7 +92,11 @@ module.exports = DatasetsItem.extend({
   },
 
   _canImportDataset: function() {
-    return ( this.user.get('remaining_byte_quota') * UploadConfig.fileTimesBigger ) >= ( this.table.get('size') || 0 )
+    var table_size = this.table.get('size') || 0;
+    return (
+        this.user.get('remaining_byte_quota') * UploadConfig.fileTimesBigger >= table_size &&
+        this.user.get('limits')['import_file_size'] > table_size
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js]: What does this change mean=

File: lib/assets/javascripts/cartodb/dashboard/views/datasets_item.jst.ejs [hunk 6]
```diff
@@ -11,6 +11,9 @@
     <% } %>
     class="iconFont iconFont-Wifi DatasetsList-itemStatus is-<%- syncStatus %>"></i>
   <% } %>
+  <% if (from_external_source) { %>
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/views/datasets_item.jst.ejs]: Let's try to use _camelCase_ instead of variables with _ separator.

File: lib/assets/javascripts/cartodb/models/synchronization.js [hunk 7]
```diff
@@ -52,6 +52,15 @@
         type_guessing: c.type_guessing
       };
 
+      if (c.type === "remote") {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/models/synchronization.js]: This needs a test.

File: lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js (hunk 4 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js]: I'll fix the conflicts if @viddo merges his branch first.

File: lib/assets/javascripts/cartodb/table/header/views/sync_info_content.jst.ejs [hunk 8]
```diff
@@ -23,11 +23,13 @@
         Error code <%- error_code %>: <%- error_message %>.
       <% } %>
 
-      <% if (state !== "syncing") { %>
-        <% if (canSync) { %>
-          <a href="#/sync-now" class="sync_now">Sync now</a>
-        <% } else { %>
-          <em class="italic light sync_now_disabled">Sync now</em>
+      <% if (!from_external_source) { %>
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/header/views/sync_info_content.jst.ejs]: ^^^ _camelCase_ thing

File: lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js [hunk 9]
```diff
@@ -49,7 +49,8 @@ module.exports = DatasetsItem.extend({
       maxTagsToShow:           3,
       canImportDataset:        this._canImportDataset(),
       rowCount:                undefined,
-      datasetSize:             undefined
+      datasetSize:             undefined,
+      from_external_source:    vis.get("synchronization") ? vis.get("synchronization").from_external_source : false
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js]: I'd move this logic in the step where we check if the synchronization model exists.

File: lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js (hunk 3 above)
Comment (Kartones) [inline lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js]: For external sources, backend forcibly assigns an interval. It can be sent but will be overriden with the one set from BE. Maybe you should then add a comment there explaining that difference.

File: lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js (hunk 5 above)
Comment (Kartones) [inline lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js]: *Do not allow to import data-library files with size bigger than user_data.limits.import_file_size*.

This is not strictly related to remote-sources, it was a faulty scenario I found, no matter your free space, if we don't check the user limits a big data-library import will fail, so how that "disabling logic" takes that into account.

File: lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js (hunk 3 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/background_polling/models/import_model.js]: Ok

File: lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js (hunk 5 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/datasets/remote_datasets_item.js]: :+1:

File: lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js (hunk 4 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/common/dialogs/map/sync_view.js]: Merged, just :telephone_receiver: me if you need :sos:

**Comment by xavijam** [comment]
Not sure about the _from_external_source_ thing, but you would have considered all the options and I trust 100%. :+1: go guys go!

File: lib/assets/javascripts/cartodb/models/synchronization.js (hunk 7 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb/models/synchronization.js]: Added.
